plugins/woocommerce/includes/class-wc-cart.php

```diff
@@ -1380,11 +1380,20 @@ public function remove_cart_item( $cart_item_key ) {
 	 * Restore a cart item.
 	 *
 	 * @param  string $cart_item_key Cart item key to restore to the cart.
+	 * @param  array  $cart_item_data Cart item data (optional, used when removed_cart_contents is not in session).
 	 * @return bool
 	 */
-	public function restore_cart_item( $cart_item_key ) {
+	public function restore_cart_item( $cart_item_key, $cart_item_data = null ) {
+		$restore_item = null;
+
 		if ( isset( $this->removed_cart_contents[ $cart_item_key ] ) ) {
-			$restore_item                                  = $this->removed_cart_contents[ $cart_item_key ];
+			$restore_item = $this->removed_cart_contents[ $cart_item_key ];
+			unset( $this->removed_cart_contents[ $cart_item_key ] );
+		} elseif ( $cart_item_data && is_array( $cart_item_data ) ) {
+			$restore_item = $cart_item_data;
+		}
+
+		if ( $restore_item ) {
 			$this->cart_contents[ $cart_item_key ]         = $restore_item;
 			$this->cart_contents[ $cart_item_key ]['data'] = wc_get_product( $restore_item['variation_id'] ? $restore_item['variation_id'] : $restore_item['product_id'] );
```

Comment:
Since we're trusting user submitted data at this point, I think we may want to run this through `$this->add_to_cart()`. I don't see that any validation is done otherwise allowing a customer to possibly manipulate the cart in ways they shouldn't be able to.

Reply:
Ya, I think this is a good call.  While I had hoped to put this off to a future PR and some of the risks exist with the current workflow that does not handle validation, the more direct input from the user leaves a slightly larger gap.

Updated in https://github.com/woocommerce/woocommerce/pull/61030/commits/7a9cff765c56c22c57db8cf863b7211e49c68fea